content/blog/2019-10-18-double-checked-locking/index.md

```diff
@@ -0,0 +1,353 @@
+            Helper h;
+            synchronized(this) {
+                h = helper;
+                if (h == null)
+                    synchronized(this) {
+                        h = new Helper();
+                    }                       
+                helper = h;
+            }
+        }
+        return helper;
+    }
+}
+
+```
+
+The intuitive idea is that the lock releasing would act as a memory barrier,
+so that `helper=h` will not be executed until the initialization 
+in the synchronized section is done.
+
+Unfortunately, the lock releasing is a one-way memory barrier on many processors.
+It only enforces that the instructions in the synchronized section 
+must be executed before lock releasing.
+The instruction `helper=h` behind the memory barrier could still be moved into 
+synchronized section and executed before the object initialization is done.
+
+## Working Solutions
+
+
+### Volatile in JDK5
+
+Java introduced "volatile" since JDK5. 
+A "volatile" makes the variable be stored and accessed in main memory.
+Every read of a volatile is from main memory.
+Every write of a volatile is to main memory.
+
+The "volatile" also provides a "happen-before" guarantee:
+
+* The reads/writes of other variables before a write to a volatile variable 
+cannot be reordered after the write to the volatile variable.
+* The reads/writes of other variables after a read from a volatile variable
+cannot be reordered before the read from the volatile variable.
+
+With this new feature, the double-checked locking issue is resolved by simply changing
+the `helper` to volatile, since it guarantees that the initialization of the
+object `new Helper()` happens before the assignment to the `helper`.
+
+```Java
+class Foo {
+    private volatile Helper helper = null;
+    public Helper getHelper() {
+        if (helper == null) {
+            synchronized(this) {    
+                if (helper == null)
+                    helper = new Helper();
+            }
+        }
+        return helper;
+    }
+}
+```
+
+However, since all read and write operations of a volatile variable need to
+access main memory, which can be very slow. 
+An improvement can be done with a local variable, to reduce number of times 
+accessing main memory.
+
+```Java
+class Foo {
+    private volatile Helper helper = null;
+    public Helper getHelper() {
+        Helper h = helper;
+        if (h == null) {
+            synchronized(this) {    
+                h = helper;
+                if (h == null) {
+                    h = new Helper();
+                    helper = h;
+                }
+            }
+        }
+        return h;
+    }
+}
+
+```
+
+In cases that the `helper` in already initialized, 
+   this optimization can reduce one volatile read by returning the local variable.
+
+### Explicit Memory Barrier
+
+The last fix with two synchronized sections does not work because releasing lock
+is an implicit "one-way" memory barrier.
```

Comment:
Do you mean the _first_ fix? I think the _last_ fix, just above, actually does work…